network-config-analyzer/MultiLayerPropertiesSet.py

```diff
@@ -36,72 +37,293 @@
-        return bool(self.methods)
+        return bool(self.methods) | \
+               bool(self.paths) | \
+               bool(self.hosts)
 
     def __eq__(self, other):
         if isinstance(other, RequestAttrs):
-            res = self.methods == other.methods
+            res = self.methods == other.methods and \
+                  self.paths == other.paths and \
+                  self.hosts == other.hosts
             return res
         return NotImplemented
 
     def __str__(self):
-        return f"request.method in {str(self.methods)}"
+        return f'request.method in {self.methods}\n' \
+               f'request.url_path {self.paths}\n' \
+               f'request.hosts {self.hosts}\n'
 
     def copy(self):
         res = RequestAttrs()
-        res.methods |= self.methods
+        res.methods = self.methods.copy()
+        res.paths = self.paths.copy()
+        res.hosts = self.hosts.copy()
         return res
 
     def __and__(self, other):
         res = RequestAttrs()
-        res.methods = self.methods & other.methods
+        res.methods = self.methods.intersection(other.methods)
+        res.paths = self.paths & other.paths
+        res.hosts = self.hosts & other.hosts
         return res
 
     def __or__(self, other):
         res = RequestAttrs()
-        res.methods = self.methods | other.methods
+        res.methods = self.methods.union(other.methods)
+        res.paths = self.paths | other.paths
+        res.hosts = self.hosts | other.hosts
         return res
 
+    def __add__(self, other):
+        return self.__or__(other)
+
     def __sub__(self, other):
         res = RequestAttrs()
-        res.methods = self.methods - other.methods
+        res.methods = self.methods.difference(other.methods)
+        res.paths = self.paths - other.paths
+        res.hosts = self.hosts - other.hosts
         return res
 
     def __iand__(self, other):
-        self.methods &= other.methods
+        self.methods.intersection_update(other.methods)
+        self.paths &= other.paths
+        self.hosts &= other.hosts
         return self
 
     def __ior__(self, other):
-        self.methods |= other.methods
+        self.methods.update(other.methods)
+        self.paths |= other.paths
+        self.hosts |= other.hosts
         return self
 
+    def __iadd__(self, other):
+        return self.__ior__(other)
+
     def __isub__(self, other):
-        self.methods -= other.methods
+        self.methods.difference_update(other.methods)
+        self.paths -= other.paths
+        self.hosts -= other.hosts
         return self
 
+    def methods_contained_in(self, other):
+        return self.methods.issubset(other.methods)
+
     def contained_in(self, other):
-        return set.issubset(self.methods, other.methods)
+        return self.methods_contained_in(other) and \
+               self.paths.contained_in(other.paths) and \
+               self.hosts.contained_in(other.hosts)
 
     def add_methods(self, methods):
```

Comment:
add comment of which methods are only for testing

Reply:
done